### ISSUE TITLE
Feature: Is permutation matrix

### ISSUE THREAD

**Issue by vprusso** [vprusso/toqito#66]
Write a function to determine whether an input matrix is a permutation matrix. A permutation matrix is defined as a matrix where there is a single "1" in each row and column of the matrix. 

The functionality for this should be created in `matrix_props_props/is_permutation.py` with corresponding unit tests found in `tests/test_matrix_props/test_is_permutation.py`. The docs should also be updated in `/docs/matrices.rst` under "Properties of Matrices" with `toqito.matrix_props.is_permutation.py`

**Comment by vanzilar** [comment]
Hi .  

 Looking for an issue for the [unitaryhack] that is tractable in the remaining time.  Looks like there are a lot of good entry ones here to dig into.  Can I try it?  If this one doesn't work out is one of the others in the list ok?  🎸 🎵  

The one's I was looking at before seemed to not be a great fit in the time, however a unitary matrix check seems very doable.

**Comment by vprusso** [comment]
Hi @vanzilar. You are definitely free to give this feature a shot. I think it should be a relatively straightforward one to implement, so have at it!

**Comment by vanzilar** [comment]
OK I got something working as far as I can tell.... had a few issues with some of the lower level libraries so I coded using python directly and cannot find a significant error.  Trying to get the Pull request in now.  I think it is good, however yes, I am open to any changes and corrections.

**Comment by vprusso** [comment]
Added some comments. Also, in addition, check the failed build, as it tells you what you need to fix to get this to pass:
https://travis-ci.com/github/vprusso/toqito/builds/227467317

**Comment by vanzilar** [comment]
Hi Vincent.

   ✨ Thanks for the feedback 🎇 

       I have also have one worry.
1.  I may have misinterpreted the definition of a permutation matrix, so I am going to go over the wikipedia text and maybe go back to my quantum and linear algebra and other texts to make sure I created the right function to start with.

And my next steps and ideas are as follows after that.
1.  I plan to keep testing until I get the test.py module working as intended (instead of my hacky version) to I can test the entire package on my local machine, as I am still running an insufficient test.

2.   I am so very grateful for travis-ci and it's tests.  It acts like this cool gatekeeper to keep the code safe. It is really really cool!

3.  My next work after that is to also try to reinstall numpy until I can get the BLAS and LAPACK to install properly.  This if it works should zero out item 1.

4.  Long term I will build this out in a VM, I rushed to run it local for the hackathon and I am realizing that is really not a good idea.... the idea of having it in some sort of VM is truly a brilliant idea when there are so many packages involved.

5.  And thanks for all those comments... I think one of the comments you made in particular, not only corrects my other errors but zeroes in on what I was going to check next aka, a missing import statement I was fearing and could very well be the one creating the travis import error.  I will be adding in each of those suggested changes.

6. I am also very interested in that formatting program called black you mentuoned and what kind of interesting changes it will make.

Thank you.

**Comment by paniash** [comment]
@vanzilar Hi! I had been facing installation issues with lapack as well. I would highly recommend reading the toqito docs page on the topic. It really helps.

Also, I think using a VM for this purpose is a bit of an overkill. Rather, it would be better to do all this in a python [virtual environment](https://wiki.archlinux.org/title/Python/Virtual_environment). I did the same and it worked! 

All the best. :-)

**Comment by vprusso** [comment]
> I may have misinterpreted the definition of a permutation matrix, so I am going to go over the wikipedia text and maybe go back to my quantum and linear algebra and other texts to make sure I created the right function to start with.

I think you are overcomplicating things here. The definition of a permutation matrix doesn't require anything super involved or deep about quantum mechanics or linear algebra. It is simply a matrix--consisting of either 1's or 0's where each row has a single 1 in it as does each column. That's it. 

> I plan to keep testing until I get the test.py module working as intended (instead of my hacky version) to I can test the entire package on my local machine, as I am still running an insufficient test.

Not exactly sure what you are referring to when you say `test.py`. Are you talking about the `pytest` module we are using for testing?

> Long term I will build this out in a VM, I rushed to run it local for the hackathon and I am realizing that is really not a good idea.... the idea of having it in some sort of VM is truly a brilliant idea when there are so many packages involved.

Not sure what you're suggesting here--I'm not sure why you would need a VM for working on a local feature. 

> Long term I will build this out in a VM, I rushed to run it local for the hackathon and I am realizing that is really not a good idea.... the idea of having it in some sort of VM is truly a brilliant idea when there are so many packages involved.

Right, that was @paniash that brought that up. The module is called `black` and this is indeed the style guideline by which the `toqito` code is formatted to. 

> And thanks for all those comments... I think one of the comments you made in particular, not only corrects my other errors but zeroes in on what I was going to check next aka, a missing import statement I was fearing and could very well be the one creating the travis import error. I will be adding in each of those suggested changes.

Yes, but make sure you can run the code yourself prior to submitting it. You would have caught this issue if you had tried to run your feature. It seems like you may have just written code without running it, which is usually a pretty bad idea. 

> I noticed several of the suggestions were from @paniash and thanking him as well... Thank you!

Yes, kudos from me as well to @paniash for adding clarity. Cheers to you both, and do let me know if I can offer any further input.

**Comment by vanzilar** [comment]
Update.  I have put in a few of the recommended changes, however not all of them yet, and I do have a few more changes to make.

**Comment by vprusso** [comment]
@vanzilar Cool. Let me know what all the changes have been implemented. Thanks for the heads up!

**Comment by vanzilar** [comment]
Thanks so much.  

It compiled once I added in the full path in the test module to the is_permutation procedure/function.  Before I was leaving off the .is_permutation to match what I saw in the other is_???????? matrix checks imports.

In addition to that,  I do not think it was necessary however I added in _autosummary a file for the is_permutation checker however I think that has no impact.  That may need to be removed.  Let me know and I will delete it if unecessary.

Next I need to review the variable naming and coding standards, add a few more comments in and make sure I get everything in toquito building and running properly on my local machines.  It's just a matter of time.  I want my code to match the style of the rest of the toquito package.

After that I wish to review the example commented code I included and run it and make sure all of that works as expected as well.  I had put in the comments quickly and did not test the embedded examples which do not run.  I just used my head to quickly check if I though the matrix examples would pass or fail for those 2 examples, so there could be a syntax error there.

One thing that really helped was looking at the commands run by the travis-ci check that is run on every build.  Examining those commands allowed me to further configure my local machine to have pytest work.

Once I get through all of these basics, I am very interested into what I can learn more about qubits and quantum computing and  to finally be able to contribute more to open source.

So I have a bit more work to do and then everything should be completed.
